File: demo/src/main/java/cn/bingoogolapple/progressbar/demo/MainActivity.java
package cn.bingoogolapple.progressbar.demo;

import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import androidx.appcompat.app.AppCompatActivity;
import android.widget.ProgressBar;

import java.lang.ref.WeakReference;

public class MainActivity extends AppCompatActivity {
    private static final int WHAT_UPDATE = 1;

    private ProgressBar mPbMainDemo1;
    private ProgressBar mPbMainDemo2;
    private ProgressBar mPbMainDemo3;
    private ProgressBar mPbMainDemo4;
    private ProgressBar mPbMainDemo5;
    private ProgressBar mPbMainDemo6;
    private ProgressBar mPbMainDemo7;

    private final UpdateProgressHandler mHandler = new UpdateProgressHandler(this);

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        mPbMainDemo1 = findViewById(R.id.pb_main_demo1);
        mPbMainDemo2 = findViewById(R.id.pb_main_demo2);
        mPbMainDemo3 = findViewById(R.id.pb_main_demo3);
        mPbMainDemo4 = findViewById(R.id.pb_main_demo4);
        mPbMainDemo5 = findViewById(R.id.pb_main_demo5);
        mPbMainDemo6 = findViewById(R.id.pb_main_demo6);
        mPbMainDemo7 = findViewById(R.id.pb_main_demo7);

        mPbMainDemo1.setMax(100);
        mPbMainDemo2.setMax(100);
        mPbMainDemo3.setMax(100);
        mPbMainDemo4.setMax(100);
        mPbMainDemo5.setMax(100);
        mPbMainDemo6.setMax(100);
        mPbMainDemo7.setMax(100);

        mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 400);
    }

    @Override
    protected void onDestroy() {
        mHandler.removeMessages(WHAT_UPDATE);
        super.onDestroy();
    }

    private void updateProgress() {
        int progress = mPbMainDemo1.getProgress();
        progress++;
        if (progress < mPbMainDemo1.getMax()) {
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 100);
        } else if (progress == mPbMainDemo1.getMax()) {
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 2000);
        } else {
            progress = 0;
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 100);
        }

        mPbMainDemo1.setProgress(progress);
        mPbMainDemo2.setProgress(progress);
        mPbMainDemo3.setProgress(progress);
        mPbMainDemo4.setProgress(progress);
        mPbMainDemo5.setProgress(progress);
        mPbMainDemo6.setProgress(progress);
        mPbMainDemo7.setProgress(progress);
    }

    public void jumpToDefault(android.view.View v) {
        startActivity(new Intent(this, DefaultActivity.class));
    }

    private static class UpdateProgressHandler extends Handler {
        private final WeakReference<MainActivity> mActivity;

        UpdateProgressHandler(MainActivity activity) {
            mActivity = new WeakReference<>(activity);
        }

        @Override
        public void handleMessage(Message msg) {
            MainActivity activity = mActivity.get();
            if (activity != null) {
                activity.updateProgress();
            }
        }
    }
}

File: demo/src/main/java/cn/bingoogolapple/progressbar/demo/MainActivity.kt
package cn.bingoogolapple.progressbar.demo

import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.support.v7.app.AppCompatActivity
import android.view.View
import kotlinx.android.synthetic.main.activity_main.*
import org.jetbrains.anko.startActivity
import java.lang.ref.WeakReference

class MainActivity : AppCompatActivity() {
    private val mHandler: UpdateProgressHandler = UpdateProgressHandler(this)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 400)
    }

    override fun onDestroy() {
        mHandler.removeMessages(WHAT_UPDATE)
        super.onDestroy()
    }

    private fun updateProgress() {
        var progress = pb_main_demo1.progress
        progress++
        if (progress < pb_main_demo1.max) {
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 100)
        } else if (progress == pb_main_demo1.max) {
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 2000)
        } else {
            progress = 0
            mHandler.sendEmptyMessageDelayed(WHAT_UPDATE, 100)
        }

        pb_main_demo1.progress = progress
        pb_main_demo2.progress = progress
        pb_main_demo3.progress = progress
        pb_main_demo4.progress = progress
        pb_main_demo5.progress = progress
        pb_main_demo6.progress = progress
        pb_main_demo7.progress = progress
    }

    /**
     * 查看系统自带 ProgressBar 支持的样式
     */
    fun jumpToDefault(v: View) {
        startActivity<DefaultActivity>()
    }

    companion object {
        private val WHAT_UPDATE = 1
    }

    private static class UpdateProgressHandler(activity: MainActivity) : Handler() {
        private val mActivity: WeakReference<MainActivity> = WeakReference<MainActivity>(activity)

        override fun handleMessage(msg: Message) {
            val activity = mActivity.get()
            activity?.apply {
                updateProgress()
            }
        }
    }
}
